File: vl2vlapi/src/translator.js
// Recursive translation of a Vega-Lite spec (plain object) into
// vega-lite-api JavaScript source.
//
// Unlike a language whose visualization wrapper library (e.g. Python's
// Altair) needs a hand-built class per nested schema type, vega-lite-api's
// code generator gives nearly every chart/channel/transform property a
// plain `.key(value)` setter method that stores whatever it's given
// verbatim (see README/ARCHITECTURE notes). That means almost every
// Vega-Lite JSON property maps 1:1 onto a method of the exact same name,
// with the parsed JSON value passed straight through as the argument --
// no merging shared properties down into composition children, no
// per-transform-type argument tables, no reserved-word renaming. The only
// property that genuinely needs special handling is `encoding`, because
// `.encode(...)` needs each argument to know its own channel name (that's
// what `vl.x(...)`/`vl.color(...)`/etc. provide).

import {formatValue} from './literals.js';
import {Chain} from './calls.js';
import {renderEncodeArgs} from './encoding.js';

const STRUCTURAL_KEYS = new Set([
  'layer', 'facet', 'spec', 'repeat', 'hconcat', 'vconcat', 'concat',
  'datasets', '$schema',
]);

class Emitter {
  constructor(includeSourcePaths = false) {
    this.lines = [];
    this.counts = new Map();
    this.includeSourcePaths = includeSourcePaths;
  }

  newVar(hint) {
    const n = (this.counts.get(hint) || 0) + 1;
    this.counts.set(hint, n);
    return n === 1 ? hint : `${hint}${n}`;
  }

  // `path` is a dotted/bracketed JSON path into the *original* spec (e.g.
  // "encoding.x", "transform[1]") identifying which part of the input this
  // one generated statement came from -- emitted as a comment directly
  // above it (opt-in via `includeSourcePaths`, off by default).
  addStmt(line, path) {
    if (this.includeSourcePaths && path) this.lines.push(`// from: ${path}`);
    this.lines.push(line);
  }
}

function sanitizeIdentifier(name, fallback) {
  return typeof name === 'string' && /^[A-Za-z_$][A-Za-z0-9_$]*$/.test(name) ? name : fallback;
}

function hoistDatasets(datasets, emitter) {
  const refs = {};
  for (const name of Object.keys(datasets)) {
    const varName = emitter.newVar(sanitizeIdentifier(name, 'dataset'));
    emitter.addStmt(`const ${varName} = ${formatValue(datasets[name])};`);
    refs[name] = varName;
  }
  return refs;
}

// Render the argument to `.data(...)`. Bare `{values: [...]}` data is
// hoisted into a `const` variable (purely for readability -- unlike some
// wrapper libraries, vega-lite-api requires no particular object identity
// for data to be shared or recognized correctly, so this is a style choice,
// not a correctness requirement).
function renderData(data, emitter, hint, path) {
  if (data === null || data === undefined) return null;
  const keys = Object.keys(data);
  if (keys.length === 1 && keys[0] === 'values') {
    const varName = emitter.newVar(hint);
    emitter.addStmt(`const ${varName} = ${formatValue(data.values)};`, path);
    return varName;
  }
  return formatValue(data);
}

// vega-lite-api renames a couple of JSON properties to avoid ambiguity with
// the constructor-style shortcuts of the same name (`mark`, `encoding`):
// `projection` -> `.project(...)`, `encoding` -> `.encode(...)` (the latter
// already gets its own dedicated handling below since it needs channel
// wrapping too).
const METHOD_RENAME = {projection: 'project'};

function legacySelectionToParam(name, def) {
  const {type, ...rest} = def;
  const select = {type: type || 'single', ...rest};
  return {name, select};
}

function applyRemaining(chain, spec, consumed, emitter, hint, path = '') {
  for (const key of Object.keys(spec)) {
    if (consumed.has(key) || key === '$schema') continue;

    if (key === 'encoding') {
      // A zero-argument `.encode()` call is a no-op in vega-lite (an empty
      // `encoding: {}` is valid) but is mishandled by vega-lite-api itself,
      // which returns `undefined` rather than `this` -- so skip the call
      // entirely rather than emit one that would break the chain.
      const args = renderEncodeArgs(spec.encoding);
      if (args.length) {
        const encPath = Object.keys(spec.encoding).map(k => `${path}encoding.${k}`).join(', ');
        chain.call('encode', args, encPath);
      }
    } else if (key === 'transform') {
      if (spec.transform.length) {
        const tPath = spec.transform.map((_, i) => `${path}transform[${i}]`).join(', ');
        chain.call('transform', spec.transform.map(t => formatValue(t)), tPath);
      }
    } else if (key === 'params') {
      if (spec.params.length) {
        const pPath = spec.params.map((_, i) => `${path}params[${i}]`).join(', ');
        chain.call('params', spec.params.map(p => formatValue(p)), pPath);
      }
    } else if (key === 'selection') {
      const names = Object.keys(spec.selection);
      const params = names.map(name => legacySelectionToParam(name, spec.selection[name]));
      if (params.length) {
        const sPath = names.map(n => `${path}selection.${n}`).join(', ');
        chain.call('params', params.map(p => formatValue(p)), sPath);
      }
    } else if (key === 'data') {
      const rendered = renderData(spec.data, emitter, `${hint}Data`, `${path}data`);
      if (rendered !== null) chain.call('data', [rendered], `${path}data`);
    } else {
      chain.call(METHOD_RENAME[key] || key, [formatValue(spec[key])], `${path}${key}`);
    }
  }
}

// A composition child's own rendered expression can start with its own
// leading `// from: ...` comment (when it has a labeled `basePath`, e.g. a
// unit view's own "mark" step) -- splicing that directly after an opening
// `(`/`vl.layer(` on the same line would put real code (the child's own
// mark/encode chain) inside that same comment's reach until the next
// newline. Still syntactically valid JS either way (the child's own code
// picks back up on its own next line, inside the still-open call), but
// confusing to read -- indenting the child onto its own fresh line avoids
// the appearance of a broken argument list.
function indentBlock(text) {
  return text
    .split('\n')
    .map(line => `  ${line}`)
    .join('\n');
}

function wrapChildExpr(childExpr, includeSourcePaths) {
  if (!includeSourcePaths) return `(${childExpr})`;
  return `(\n${indentBlock(childExpr)}\n)`;
}

function joinChildArgs(children, includeSourcePaths) {
  if (!includeSourcePaths) return children.join(', ');
  return `\n${children.map(indentBlock).join(',\n')}\n`;
}

function translateUnit(spec, emitter, hint, path = '') {
  const chain = new Chain(`vl.mark(${formatValue(spec.mark)})`, {
    basePath: path ? `${path}mark` : undefined,
    includeSourcePaths: emitter.includeSourcePaths,
  });
  applyRemaining(chain, spec, new Set(['mark']), emitter, hint, path);
  return chain.toString();
}

function translateLayer(spec, emitter, hint, path = '') {
  const childHint = hint === 'chart' ? 'layer' : hint;
  const children = spec.layer.map(
    (child, i) => translateSpec(child, emitter, `${childHint}${i + 1}`, `${path}layer[${i}].`)
  );
  const chain = new Chain(`vl.layer(${joinChildArgs(children, emitter.includeSourcePaths)})`, {
    basePath: path ? `${path}layer` : undefined,
    includeSourcePaths: emitter.includeSourcePaths,
  });
  applyRemaining(chain, spec, new Set(['layer']), emitter, hint, path);
  return chain.toString();
}

function translateMulti(spec, emitter, hint, key, fn, path = '') {
  const childHint = hint === 'chart' ? key : hint;
  const children = spec[key].map(
    (child, i) => translateSpec(child, emitter, `${childHint}${i + 1}`, `${path}${key}[${i}].`)
  );
  const chain = new Chain(`vl.${fn}(${joinChildArgs(children, emitter.includeSourcePaths)})`, {
    basePath: path ? `${path}${key}` : undefined,
    includeSourcePaths: emitter.includeSourcePaths,
  });
  applyRemaining(chain, spec, new Set([key]), emitter, hint, path);
  return chain.toString();
}

function translateFacet(spec, emitter, hint, path = '') {
  const childHint = hint === 'chart' ? 'view' : `${hint}View`;
  const childExpr = translateSpec(spec.spec, emitter, childHint, `${path}spec.`);
  const chain = new Chain(
    `${wrapChildExpr(childExpr, emitter.includeSourcePaths)}.facet(${formatValue(spec.facet)})`,
    {basePath: path ? `${path}facet` : undefined, includeSourcePaths: emitter.includeSourcePaths}
  );
  applyRemaining(chain, spec, new Set(['facet', 'spec']), emitter, hint, path);
  return chain.toString();
}

function translateRepeat(spec, emitter, hint, path = '') {
  const childHint = hint === 'chart' ? 'view' : `${hint}View`;
  const childExpr = translateSpec(spec.spec, emitter, childHint, `${path}spec.`);
  const chain = new Chain(
    `${wrapChildExpr(childExpr, emitter.includeSourcePaths)}.repeat(${formatValue(spec.repeat)})`,
    {basePath: path ? `${path}repeat` : undefined, includeSourcePaths: emitter.includeSourcePaths}
  );
  applyRemaining(chain, spec, new Set(['repeat', 'spec']), emitter, hint, path);
  return chain.toString();
}

export function translateSpec(spec, emitter, hint = 'chart', path = '') {
  const rest = {...spec};
  delete rest.$schema;

  if ('layer' in rest) return translateLayer(rest, emitter, hint, path);
  if ('facet' in rest && 'spec' in rest) return translateFacet(rest, emitter, hint, path);
  if ('repeat' in rest && 'spec' in rest) return translateRepeat(rest, emitter, hint, path);
  if ('hconcat' in rest) return translateMulti(rest, emitter, hint, 'hconcat', 'hconcat', path);
  if ('vconcat' in rest) return translateMulti(rest, emitter, hint, 'vconcat', 'vconcat', path);
  if ('concat' in rest) return translateMulti(rest, emitter, hint, 'concat', 'concat', path);
  return translateUnit(rest, emitter, hint, path);
}

// `includeSourcePaths` (default `false`): when `true`, each generated
// statement/chain-step is preceded by a `// from: <json path>` comment
// naming the part of the *input* spec it was translated from (e.g.
// `// from: encoding.x`, `// from: transform[0]`) -- useful for tracing
// generated code back to the spec that produced it, at the cost of a much
// noisier script (every chain is forced one step per line so the comments
// have somewhere to go).
export function specToCode(spec, {chartVar = 'chart', includeSourcePaths = false} = {}) {
  const emitter = new Emitter(includeSourcePaths);
  const root = {...spec};
  delete root.$schema;
  const datasets = root.datasets;
  delete root.datasets;

  let datasetRefs = null;
  if (datasets && Object.keys(datasets).length) {
    datasetRefs = hoistDatasets(datasets, emitter);
  }

  let expr = translateSpec(root, emitter, chartVar);
  if (datasetRefs) {
    // Build the `.datasets({...})` argument by hand: its values must be raw
    // identifier tokens (the hoisted `const` variables), not JSON literals,
    // so `formatValue` (which only knows how to render JSON-shaped data)
    // doesn't apply here.
    const pairs = Object.entries(datasetRefs).map(
      ([name, varName]) => `${JSON.stringify(name)}: ${varName}`
    );
    const chain = new Chain(`(${expr})`, {includeSourcePaths});
    chain.call('datasets', [`{${pairs.join(', ')}}`], 'datasets');
    expr = chain.toString();
  }

  // A provenance header: this file is machine-generated from a Vega-Lite
  // spec, not hand-written -- re-run the translator (with the same
  // arguments shown here) after the source spec changes, rather than
  // hand-editing this output and losing that round-trip.
  const headerComment =
    `// Generated by vl2vlapi.vegaLiteToVlApiCode(spec, {chartVar: ${JSON.stringify(chartVar)}, ` +
    `includeSourcePaths: ${includeSourcePaths}})`;
  const lines = [headerComment, "import * as vl from 'vega-lite-api';", ''];
  lines.push(...emitter.lines);
  if (emitter.lines.length) lines.push('');
  lines.push(`const ${chartVar} = ${expr};`);
  lines.push('');
  lines.push(`export default ${chartVar};`);
  lines.push('');
  return lines.join('\n');
}

File: vl2altair/translator.py
"""Recursive translation of a Vega-Lite spec (dict) into Altair Python source."""

from __future__ import annotations

import keyword

from .calls import render_call_kv, render_kwargs
from .encoding import render_encoding_kwargs
from .literals import format_value, try_black_format
from .params import render_legacy_selection, render_param
from .transforms import render_transform_call

# Keys that describe *structure* (composition/spec-nesting) rather than a
# property that can be passed through Chart.properties(); these should never
# leak into a generic ``.properties(**leftover)`` call.
STRUCTURAL_KEYS = {
    "layer",
    "facet",
    "spec",
    "repeat",
    "hconcat",
    "vconcat",
    "concat",
    "$schema",
    "datasets",
}


class Emitter:
    def __init__(self) -> None:
        self.lines: list[str] = []
        self.dataset_vars: dict[str, str] = {}
        self._counts: dict[str, int] = {}
        self._data_var_cache: dict[int, str] = {}

    def new_var(self, hint: str) -> str:
        n = self._counts.get(hint, 0) + 1
        self._counts[hint] = n
        return hint if n == 1 else f"{hint}_{n}"

    def add_stmt(self, line: str) -> None:
        self.lines.append(line)


def _sanitize_identifier(name: object, fallback: str) -> str:
    if isinstance(name, str) and name.isidentifier() and not keyword.iskeyword(name):
        return name
    return fallback


def _hoist_datasets(datasets: dict, emitter: Emitter) -> None:
    for name, values in datasets.items():
        hint = _sanitize_identifier(name, "dataset")
        var = emitter.new_var(hint)
        emitter.add_stmt(f"{var} = {format_value(values)}")
        emitter.dataset_vars[name] = var


def _render_data(data: object, emitter: Emitter, hint: str) -> str | None:
    """Render a Vega-Lite ``data`` value into a Python expression.

    Every dict-shaped data value is hoisted into its own variable rather than
    inlined, and repeated calls for the *same* dict object (by identity) reuse
    the same variable. This matters because Altair's own logic for
    consolidating identical data across layer/concat/facet children
    (``_combine_subchart_data``) checks object identity, not equality -- and
    since ``_merge_down`` shares (rather than copies) a wrapper's ``data``
    dict across its children, giving each occurrence the same Python object
    here lets that consolidation succeed instead of silently failing.
    """
    if data is None:
        return None
    if isinstance(data, dict):
        cached = emitter._data_var_cache.get(id(data))
        if cached is not None:
            return cached
        if set(data.keys()) == {"name"} and data["name"] in emitter.dataset_vars:
            rendered = f'{{"values": {emitter.dataset_vars[data["name"]]}}}'
        else:
            rendered = format_value(data)
        var = emitter.new_var(hint)
        emitter.add_stmt(f"{var} = {rendered}")
        emitter._data_var_cache[id(data)] = var
        return var
    return format_value(data)


def _merge_down(child: dict, wrapper: dict, merge_encoding: bool) -> dict:
    """Push wrapper-level ``data``/``transform`` (and, for layers, ``encoding``)
    down into a composition child so each child can be translated as if it
    were fully self-contained."""
    child = dict(child)
    if "data" not in child and "data" in wrapper:
        child["data"] = wrapper["data"]
    if "transform" in wrapper:
        child["transform"] = list(wrapper["transform"]) + list(child.get("transform", []))
    if merge_encoding and "encoding" in wrapper:
        merged_enc = dict(wrapper["encoding"])
        merged_enc.update(child.get("encoding", {}))
        child["encoding"] = merged_enc
    return child


def _apply_common(varname: str, spec: dict, emitter: Emitter, consumed: set) -> None:
    """Apply transform/params/resolve/projection/config, and route anything
    left over through a generic ``.properties(**leftover)`` call."""
    if "transform" in spec:
        for t in spec["transform"]:
            call = render_transform_call(t)
            if call is None:
                emitter.add_stmt(
                    f"{varname}.transform = list({varname}.transform "
                    f"if {varname}.transform is not alt.Undefined else []) + "
                    f"[{format_value(t)}]"
                )
            else:
                emitter.add_stmt(f"{varname} = {varname}.{call}")
        consumed.add("transform")

    param_vars = []
    if "params" in spec:
        for p in spec["params"]:
            pv = emitter.new_var("param")
            emitter.add_stmt(f"{pv} = {render_param(p)}")
            param_vars.append(pv)
        consumed.add("params")
    if "selection" in spec:
        for name, definition in spec["selection"].items():
            pv = emitter.new_var("param")
            emitter.add_stmt(f"{pv} = {render_legacy_selection(name, definition)}")
            param_vars.append(pv)
        consumed.add("selection")
    if param_vars:
        emitter.add_stmt(f"{varname} = {varname}.add_params({', '.join(param_vars)})")

    if "resolve" in spec:
        resolve = spec["resolve"]
        for kind in ("scale", "axis", "legend"):
            if kind in resolve:
                emitter.add_stmt(
                    f"{varname} = {varname}.resolve_{kind}({render_kwargs(resolve[kind])})"
                )
        consumed.add("resolve")

    if "projection" in spec:
        emitter.add_stmt(f"{varname} = {varname}.project({render_kwargs(spec['projection'])})")
        consumed.add("projection")

    if "config" in spec:
        emitter.add_stmt(f"{varname} = {varname}.configure({render_kwargs(spec['config'])})")
        consumed.add("config")

    leftover = {k: v for k, v in spec.items() if k not in consumed and k not in STRUCTURAL_KEYS}
    if leftover:
        emitter.add_stmt(f"{varname} = {varname}.properties({render_kwargs(leftover)})")


def _translate_unit(spec: dict, emitter: Emitter, hint: str) -> str:
    varname = emitter.new_var(hint)
    data_hint = "source" if hint == "chart" else f"{hint}_data"
    data_expr = _render_data(spec.get("data"), emitter, data_hint)
    emitter.add_stmt(f"{varname} = alt.Chart({data_expr or ''})")
    consumed = {"data"}

    if "mark" in spec:
        mark = spec["mark"]
        if isinstance(mark, str):
            mtype, mkwargs = mark, {}
        else:
            mark = dict(mark)
            mtype = mark.pop("type")
            mkwargs = mark
        emitter.add_stmt(f"{varname} = {varname}.mark_{mtype}({render_kwargs(mkwargs)})")
        consumed.add("mark")

    if "encoding" in spec:
        enc_kwargs = render_encoding_kwargs(spec["encoding"])
        emitter.add_stmt(render_call_kv(f"{varname} = {varname}.encode", list(enc_kwargs.items())))
        consumed.add("encoding")

    _apply_common(varname, spec, emitter, consumed)
    return varname


def _translate_layer(spec: dict, emitter: Emitter, hint: str) -> str:
    wrapper = spec
    base_hint = "layer" if hint == "chart" else hint
    children_vars = [
        translate_spec(_merge_down(child, wrapper, merge_encoding=True), emitter, f"{base_hint}{i}")
        for i, child in enumerate(wrapper["layer"], start=1)
    ]
    varname = emitter.new_var(hint)
    emitter.add_stmt(f"{varname} = alt.layer({', '.join(children_vars)})")
    consumed = {"layer", "data", "transform", "encoding"}
    _apply_common(varname, wrapper, emitter, consumed)
    return varname


def _translate_multi(spec: dict, emitter: Emitter, hint: str, key: str, func: str) -> str:
    wrapper = spec
    base_hint = key if hint == "chart" else hint
    children_vars = [
        translate_spec(_merge_down(child, wrapper, merge_encoding=False), emitter, f"{base_hint}{i}")
        for i, child in enumerate(wrapper[key], start=1)
    ]
    varname = emitter.new_var(hint)
    emitter.add_stmt(f"{varname} = {func}({', '.join(children_vars)})")
    consumed = {key, "data", "transform"}
    _apply_common(varname, wrapper, emitter, consumed)
    return varname


def _translate_facet(spec: dict, emitter: Emitter, hint: str) -> str:
    wrapper = spec
    child_hint = "view" if hint == "chart" else f"{hint}_view"
    child = _merge_down(wrapper["spec"], wrapper, merge_encoding=False)
    child_var = translate_spec(child, emitter, child_hint)

    facet_val = wrapper["facet"]
    fkwargs: dict = {}
    if isinstance(facet_val, dict) and ("row" in facet_val or "column" in facet_val):
        if "row" in facet_val:
            fkwargs["row"] = facet_val["row"]
        if "column" in facet_val:
            fkwargs["column"] = facet_val["column"]
    else:
        fkwargs["facet"] = facet_val
    if "columns" in wrapper:
        fkwargs["columns"] = wrapper["columns"]

    varname = emitter.new_var(hint)
    emitter.add_stmt(f"{varname} = {child_var}.facet({render_kwargs(fkwargs)})")
    consumed = {"facet", "spec", "data", "transform", "columns"}
    _apply_common(varname, wrapper, emitter, consumed)
    return varname


def _translate_repeat(spec: dict, emitter: Emitter, hint: str) -> str:
    wrapper = spec
    child_hint = "view" if hint == "chart" else f"{hint}_view"
    child = _merge_down(wrapper["spec"], wrapper, merge_encoding=False)
    child_var = translate_spec(child, emitter, child_hint)

    repeat_val = wrapper["repeat"]
    rkwargs: dict = {}
    if isinstance(repeat_val, list):
        rkwargs["repeat"] = repeat_val
    else:
        for k in ("row", "column", "layer"):
            if k in repeat_val:
                rkwargs[k] = repeat_val[k]
    if "columns" in wrapper:
        rkwargs["columns"] = wrapper["columns"]

    varname = emitter.new_var(hint)
    emitter.add_stmt(f"{varname} = {child_var}.repeat({render_kwargs(rkwargs)})")
    consumed = {"repeat", "spec", "data", "transform", "columns"}
    _apply_common(varname, wrapper, emitter, consumed)
    return varname


def translate_spec(spec: dict, emitter: Emitter, hint: str = "chart") -> str:
    spec = dict(spec)
    spec.pop("$schema", None)

    if "layer" in spec:
        return _translate_layer(spec, emitter, hint)
    if "facet" in spec and "spec" in spec:
        return _translate_facet(spec, emitter, hint)
    if "repeat" in spec and "spec" in spec:
        return _translate_repeat(spec, emitter, hint)
    if "hconcat" in spec:
        return _translate_multi(spec, emitter, hint, "hconcat", "alt.hconcat")
    if "vconcat" in spec:
        return _translate_multi(spec, emitter, hint, "vconcat", "alt.vconcat")
    if "concat" in spec:
        return _translate_multi(spec, emitter, hint, "concat", "alt.concat")
    return _translate_unit(spec, emitter, hint)


def spec_to_code(spec: dict, chart_var: str = "chart", format_with_black: bool = True) -> str:
    """Translate a full Vega-Lite JSON spec (as a Python dict) into a
    standalone Altair Python script (as a string)."""
    emitter = Emitter()
    root = dict(spec)
    root.pop("$schema", None)
    datasets = root.pop("datasets", None)
    if datasets:
        _hoist_datasets(datasets, emitter)

    varname = translate_spec(root, emitter, chart_var)

    # A provenance header: this file is machine-generated from a Vega-Lite
    # spec, not hand-written -- re-run the translator (with the same
    # arguments shown here) after the source spec changes, rather than
    # hand-editing this output and losing that round-trip.
    header_comment = (
        f"# Generated by vl2altair.vegalite_to_altair_code(spec, "
        f"chart_var={chart_var!r}, format_with_black={format_with_black!r})"
    )
    lines = [header_comment, "import altair as alt", ""]
    lines.extend(emitter.lines)
    if varname != chart_var:
        lines.append(f"{chart_var} = {varname}")
    lines.append("")
    lines.append(chart_var)

    code = "\n".join(lines) + "\n"
    if format_with_black:
        code = try_black_format(code)
    return code
